crypto/asn1/a_time.c

```diff
@@ -509,14 +511,60 @@ int ossl_asn1_time_print_ex(BIO *bp, const ASN1_TIME *tm)
         }
 
         return BIO_printf(bp, "%s %2d %02d:%02d:%02d%.*s %d%s",
-                          _asn1_mon[stm.tm_mon], stm.tm_mday, stm.tm_hour,
-                          stm.tm_min, stm.tm_sec, f_len, f, stm.tm_year + 1900,
-                          (gmt ? " GMT" : "")) > 0;
+                    _asn1_mon[stm.tm_mon], stm.tm_mday, stm.tm_hour,
+                    stm.tm_min, stm.tm_sec, f_len, f, stm.tm_year + 1900,
+                    (gmt ? " GMT" : "")) > 0;
     } else {
         return BIO_printf(bp, "%s %2d %02d:%02d:%02d %d%s",
-                          _asn1_mon[stm.tm_mon], stm.tm_mday, stm.tm_hour,
-                          stm.tm_min, stm.tm_sec, stm.tm_year + 1900,
-                          (gmt ? " GMT" : "")) > 0;
+                    _asn1_mon[stm.tm_mon], stm.tm_mday, stm.tm_hour,
+                    stm.tm_min, stm.tm_sec, stm.tm_year + 1900,
+                    (gmt ? " GMT" : "")) > 0;
+    }
+}
+
+/* returns 0 on BIO write error, else -1 in case of parse failure, else 1 */
+int ossl_asn1_time_print_ex_iso(BIO *bp, const ASN1_TIME *tm)
+{
+    char *v;
+    int gmt = 0, l;
+    struct tm stm;
+    const char upper_z = 0x5A, period = 0x2E;
+
+    /* ossl_asn1_time_to_tm will check the time type */
+    if (!ossl_asn1_time_to_tm(&stm, tm))
+        return BIO_write(bp, "Bad time value", 14) ? -1 : 0;
+
+    l = tm->length;
+    v = (char *)tm->data;
+    if (v[l - 1] == upper_z)
```

Comment:
l=0 would be  bad.

Reply:
So this is kind of a hack because I copy/pasted to duplicate `ossl_asn1_time_print_ex` into `ossl_asn1_time_print_ex_rfc` and `ossl_asn1_time_print_ex_iso`, while only modifying the `BIO_printf()` arguments.

I can refactor this back into one function to repeat less code, but if you think this logic may be flawed we can investigate that as well.